### PR TITLE
[Bugfix][Rust Frontend] Validate out-of-vocab token ids in request params

### PR DESCRIPTION
**Pull request by Sunt-ing**
## Purpose

With `VLLM_USE_RUST_FRONTEND=1`, `allowed_token_ids`, `logit_bias` keys, and `/v1/completions` token-id `prompt`s are forwarded to the engine unchanged. An id at or beyond the model vocabulary makes the engine index the embedding table or the logits vector out of range, triggering a **CUDA device-side assert that aborts EngineCore and takes down the whole server** — a single well-formed request is a DoS. The Python frontend rejects these in the serving layer with HTTP 400; the Rust frontend did not.

This adds the same per-field range validation, each bound matching Python and the tensor the engine actually indexes:

- **Token-id `prompt`s and `logit_bias` keys** → **model vocabulary** (the embedding/logits rows).
- **`allowed_token_ids`** → **`len(tokenizer)`** (Python's `_validate_allowed_token_ids` rule), recovered as the base vocab plus `tokenizer_config.json`'s `added_tokens_decoder` special ids — the plain `vocab_size()` misses those and would wrongly reject the `[base_vocab, len(tokenizer))` band Python accepts.

## Implementation

- Add `Tokenizer::vocab_size` (HF/fastokens + tiktoken) and `ModelConfig::vocab_size` (nested `text_config` fallback), threaded through the text/chat/app-state layers.
- Recover `len(tokenizer)` for `allowed_token_ids` by also counting `added_tokens_decoder` ids.
- Add `validate_token_id_ranges` in each route's `validate` module, run before lowering; an unknown bound is skipped, not rejected.

## Test

`cargo test -p vllm-server -p vllm-tokenizer -p vllm-text`, `cargo clippy --all-targets -- -D warnings`, and `cargo fmt --check` are clean. End-to-end on Qwen3.5-4B (`VLLM_USE_RUST_FRONTEND=1`), this PR plus the companion detokenizer PR matches the Python frontend on every id:

| request | stock Rust | this PR + companion | Python |
|---|---|---|---|
| `logit_bias` 248100 | 200 | **200** | 200 |
| `allowed_token_ids` 248073 (special, < len(tok)) | 500 | **200** | 200 |
| `allowed_token_ids` 248100 (>= len(tok)) | 500 | **400** | 400 |
| token-id `prompt` [248100] (< model vocab) | 500 | **200** | 200 |
| token-id `prompt` [248320] (>= model vocab) | **500, then DEAD** | **400, healthy** | 400 |
| normal chat request | 200 | 200 | 200 |

The stock frontend faults the engine (`/health` goes DEAD) on ids `>= model vocab`; with the fix they get a clean 400 and the server stays healthy, zero panics.

<details>
<summary>Cross-tokenizer validation — four tokenizer families + the multimodal edge</summary>

Because the correct bound per field depends on the tokenizer-vs-model-vocab relationship, the bounds were checked against Python's actual code (`_validate_logit_bias` → `model_config.get_vocab_size()`; `_validate_allowed_token_ids` → `len(tokenizer)`; `input_processor.py` prompt bound → `max(tokenizer.max_token_id, model_vocab_size - 1)`) on models spanning all three relationships:

| model | `len(tokenizer)` | model vocab | relationship | Rust `tokenizer_vocab_size` | Rust `model_vocab_size` |
|---|---|---|---|---|---|
| TinyLlama-1.1B | 32000 | 32000 | equal | 32000 | 32000 |
| Qwen3.5-4B | 248077 | 248320 | model > tokenizer | 248077 | 248320 |
| Phi-3-mini | 32011 | 32064 | model > tokenizer | 32011 | 32064 |
| gemma-3-1b-it | 262145 | 262144 | tokenizer > model (MM placeholder) | 262145 | 262144 |

Rust computes `tokenizer_vocab_size = max(get_vocab_size(true), tokenizer_config max_added_token_id + 1)`, equal to Python `len(tokenizer)` on every contiguous-id tokenizer (and robustly recovering the `tokenizer_config` special ids a plain `get_vocab_size` misses — the 248070 vs 248077 case). `model_vocab_size` reads `config.json` `vocab_size` with a `text_config` fallback, matching `model_config.get_vocab_size()`.

**Phi-3-mini end-to-end (model 32064 > len(tok) 32011):** `prompt[32063]`→200, `prompt[32064]`→**400, engine healthy**, `allowed[32010]`→200, `allowed[32011]`→400, `logit_bias[32064]`→400, normal→200.

**gemma-3-1b, Rust-vs-Python differential (len(tok) 262145 > embedding rows 262144):** identical on every probe, including the residual engine fault on the placeholder id (see scope note):

| request | this PR (Rust) | Python |
|---|---|---|
| `prompt[262145]` / `allowed[262145]` / `logit_bias[262144]` | 400, healthy | 400, healthy |
| `prompt[262144]` (placeholder id, >= embed rows, < len(tok)) | **500, engine fault** | **500, engine fault** |

</details>

## Scope

- **Front-end parity patch, not engine hardening.** It mirrors Python's serving-layer checks at the entry; it does not stop an OOB id that reaches a GPU index op by some other path.
- **Residual engine-layer DoS on multimodal models (matches Python).** When `len(tokenizer) >` the embedding rows (e.g. gemma-3's placeholder id 262144), a `prompt`/`allowed_token_ids` in `[embed_rows, len(tokenizer))` is admitted by **both** frontends and faults EngineCore. The serving layer admits it on purpose so legitimate multimodal placeholders pass; closing it needs an engine-layer guard, the same category as `stop_token_ids` (vllm-project/vllm#42614).
- **Two vectors out of scope.** `top_logprobs`/`prompt_logprobs` are the same DoS class but need `max_logprobs` wiring (a separate semantic change); the surface is not fully safe after this PR alone.
- **Companion PR.** Full parity on the `[tokenizer_vocab, model_vocab)` band also needs `fix/rust-frontend-detok-oob-prompt-token`; orthogonal and independently testable.

AI assistance was used to investigate, reproduce, and draft this change; the author reviewed the diff and validation output.

cc @BugenZhao

### CODE REVIEW

File: rust/src/text/src/backend/hf/config.rs [hunk 1]
```diff
@@ -19,6 +21,18 @@ pub struct HfTokenizerConfig {
...
+}
+
+impl HfTokenizerConfig {
+    /// Highest token id declared in `added_tokens_decoder`, used to recover the
+    /// full tokenizer vocabulary size (Python `len(tokenizer)`), which counts
+    /// special tokens that the base `tokenizer.json` vocab omits.
+    pub fn max_added_token_id(&self) -> Option<u32> {
+        self.added_tokens_decoder.keys().next_back().copied()
+    }
 }
```
Comment (BugenZhao) [inline rust/src/text/src/backend/hf/config.rs]: I think after https://github.com/vllm-project/vllm/pull/44683 gets merged we can clean this and all related stuff up...

File: rust/src/text/src/backend/hf/mod.rs [hunk 2]
```diff
@@ -100,6 +109,14 @@ impl TextBackend for HfTextBackend {
         self.model_config.is_moe()
     }
 
+    fn model_vocab_size(&self) -> Option<usize> {
+        self.model_config.vocab_size().map(|v| v as usize)
+    }
+
+    fn tokenizer_vocab_size(&self) -> Option<usize> {
+        self.tokenizer_vocab_size
+    }
```
Comment (BugenZhao) [inline rust/src/text/src/backend/hf/mod.rs]: ...and then remove this method from the trait, or provide a default impl to `self.tokenizer().vocab_size()`.

File: rust/src/tokenizer/src/tiktoken.rs [hunk 3]
```diff
@@ -596,6 +603,17 @@ mod tests {
     /// rather than returning an error for incomplete multi-byte sequences.
     /// This is critical for streaming decode — `DecodeStream` relies on
     /// `\u{FFFD}` to detect incomplete characters.
+    #[test]
```
Comment (BugenZhao) [inline rust/src/tokenizer/src/tiktoken.rs]: The docstring seems to be placed on wrong function.

File: rust/src/tokenizer/src/lib.rs [hunk 4]
```diff
@@ -34,6 +34,11 @@ pub trait Tokenizer: Send + Sync {
         None
     }
 
+    /// Return the vocabulary size, if the backend can report it, else `None`.
+    fn vocab_size(&self) -> Option<usize> {
+        None
+    }
+
```
Comment (BugenZhao) [inline rust/src/tokenizer/src/lib.rs]: I think all tokenizers should be able to report this? If this is the case, we can remove `Option` and by default return `usize::MAX` just for testing purposes.

File: rust/src/server/src/routes/openai/completions/validate.rs [hunk 5]
```diff
@@ -104,13 +104,87 @@ pub(super) fn validate_request_compat(
     Ok(())
 }
 
+/// Reject out-of-vocab token ids, mirroring the Python frontend. Token-id prompts
+/// and `logit_bias` keys are bounded by the model vocabulary; `allowed_token_ids`
+/// by the tokenizer vocabulary (the Python serving-layer rule). Each bound is
+/// skipped when its size is unknown.
+pub(super) fn validate_token_id_ranges(
```
Comment (BugenZhao) [inline rust/src/server/src/routes/openai/completions/validate.rs:115]: This seems to be duplicated a lot with the one under `chat_completion`. Can you extract a function for each independent validation so that they can be reused?

**Review by BugenZhao — COMMENTED** [review]
Thanks!

**Comment by Sunt-ing** [comment]
@BugenZhao Thanks! Addressed all five (rebased so #44683 is in):

- Dropped the hand-rolled `added_tokens_decoder` / `max_added_token_id`; rely on `tokenizer.vocab_size()`.
- `tokenizer_vocab_size` is now a trait default (`self.tokenizer().vocab_size()`), no per-backend impl.
- `Tokenizer::vocab_size` returns `usize` (default `usize::MAX`), implemented for the hf/tiktoken/tekken backends.
- Moved the tiktoken test so the docstring reattaches to its function.
- Extracted per-field validators into `utils::token_ids`, shared by both routes (the token-id prompt bound follows the Python input processor's `max(tokenizer.max_token_id, model_vocab_size - 1)`).

**Review by BugenZhao — APPROVED** [review]
